presto-main/src/main/java/com/facebook/presto/type/EnumOperators.java

```diff
@@ -0,0 +1,349 @@
+    {
+        if (toType instanceof LongEnumType) {
+            if (INTEGER_TYPES.contains(fromType)) {
+                return buildCastToEnum(fromType, toType, "integerEnumValueLookup");
+            }
+            if (StandardTypes.VARCHAR.equals(fromType.getBase())) {
+                return buildCastToEnum(fromType, toType, "integerEnumKeyLookup");
+            }
+        }
+        if (toType instanceof VarcharEnumType) {
+            if (StandardTypes.VARCHAR.equals(fromType.getBase())) {
+                return buildCastToEnum(fromType, toType, "stringEnumKeyOrValueLookup");
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<SqlFunction> makeCastFromEnum(EnumType fromType, TypeSignature toType)
+    {
+        if (fromType instanceof LongEnumType) {
+            if (BigintType.BIGINT.getTypeSignature().equals(toType)) {
+                return buildCastFromEnum(fromType, toType, "integerEnumAsPrimitive");
+            }
+        }
+        if (fromType instanceof VarcharEnumType) {
+            if (StandardTypes.VARCHAR.equals(toType.getBase())) {
+                return buildCastFromEnum(fromType, toType, "stringEnumAsPrimitive");
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<SqlFunction> buildCastToEnum(TypeSignature fromType, EnumType toType, String methodName)
+    {
+        return buildFunction(
+                CAST,
+                ImmutableList.of(fromType),
+                toType.getTypeSignature(),
+                false,
+                methodName,
+                ImmutableList.of(toType));
+    }
+
+    private static Optional<SqlFunction> buildCastFromEnum(EnumType fromType, TypeSignature toType, String methodName)
+    {
+        return buildFunction(
+                CAST,
+                ImmutableList.of(fromType.getTypeSignature()),
+                toType,
+                false,
+                methodName,
+                ImmutableList.of());
+    }
+
+    private static Optional<SqlFunction> buildFunction(OperatorType operatorType, List<TypeSignature> argTypes, TypeSignature returnType,
+            boolean nullableResult, String methodName, List<Object> extraParams)
+    {
+        return buildFunction(operatorType, argTypes, returnType, nullableResult, methodName, extraParams, null);
+    }
+
+    private static Optional<SqlFunction> buildFunction(OperatorType operatorType, List<TypeSignature> argTypes, TypeSignature returnType,
+            boolean nullableResult, String methodName, List<Object> extraParams, BuiltInScalarFunctionImplementation.ArgumentProperty[] argumentProperties)
+    {
+        Signature signature = SignatureBuilder.builder()
+                .kind(SCALAR)
+                .operatorType(operatorType)
+                .argumentTypes(argTypes)
+                .returnType(returnType)
+                .build();
+
+        return Optional.of(SqlScalarFunction.builder(EnumOperators.class)
```

Comment:
I don't think you need `PolymorphicScalarFunction` here. This is for special cases where there are multiple physical implementation of the same logical function (the same function signature has multiple java implementations). I don't think you need that for enum operator functions. So I think you can just create an abstract `EnumSqlScalarFunction` and override `specialize` to provide the implementation.